pkg/handler/common/cluster_test.go

```diff
@@ -0,0 +1,218 @@
+		Datacenter     *kubermaticv1.Datacenter
+		Cluster        *kubermaticv1.Cluster
+		ExpectedStatus apiv1.ExternalCCMStatus
+	}{
+		{
+			Name: "scenario 1: CCM migration not needed since the beginning",
+			Datacenter: &kubermaticv1.Datacenter{
+				Spec: kubermaticv1.DatacenterSpec{Openstack: &kubermaticv1.DatacenterSpecOpenstack{}},
+			},
+			Cluster: &kubermaticv1.Cluster{
+				Spec: kubermaticv1.ClusterSpec{
+					Cloud: kubermaticv1.CloudSpec{
+						Openstack: &kubermaticv1.OpenstackCloudSpec{},
+					},
+					Features: map[string]bool{
+						kubermaticv1.ClusterFeatureExternalCloudProvider: true,
+					},
+					Version: semver.Semver{
+						Version: version,
+					},
+				},
+			},
+			ExpectedStatus: apiv1.ExternalCCMStatus{
+				ExternalCCM:          true,
+				ExternalCCMMigration: apiv1.ExternalCCMMigrationNotNeeded,
+			},
+		},
+		{
+			Name: "scenario 2: CCM migration not needed because cluster has already migrated",
+			Datacenter: &kubermaticv1.Datacenter{
+				Spec: kubermaticv1.DatacenterSpec{Openstack: &kubermaticv1.DatacenterSpecOpenstack{}},
+			},
+			Cluster: &kubermaticv1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{
+						kubermaticv1.CCMMigrationNeededAnnotation: "",
+						kubermaticv1.CSIMigrationNeededAnnotation: "",
+					},
+				},
+				Spec: kubermaticv1.ClusterSpec{
+					Cloud: kubermaticv1.CloudSpec{
+						Openstack: &kubermaticv1.OpenstackCloudSpec{},
+					},
+					Features: map[string]bool{
+						kubermaticv1.ClusterFeatureExternalCloudProvider: true,
+					},
+					Version: semver.Semver{
+						Version: version,
+					},
+				},
+				Status: kubermaticv1.ClusterStatus{
+					Conditions: []kubermaticv1.ClusterCondition{
+						{
+							Type:   kubermaticv1.ClusterConditionCSIKubeletMigrationCompleted,
+							Status: corev1.ConditionTrue,
+						},
+					},
+				},
+			},
+			ExpectedStatus: apiv1.ExternalCCMStatus{
+				ExternalCCM:          true,
+				ExternalCCMMigration: apiv1.ExternalCCMMigrationNotNeeded,
+			},
+		},
+		{
+			Name: "scenario 3: CCM migration supported",
+			Datacenter: &kubermaticv1.Datacenter{
+				Spec: kubermaticv1.DatacenterSpec{Openstack: &kubermaticv1.DatacenterSpecOpenstack{}},
+			},
+			Cluster: &kubermaticv1.Cluster{
+				Spec: kubermaticv1.ClusterSpec{
+					Cloud: kubermaticv1.CloudSpec{
+						Openstack: &kubermaticv1.OpenstackCloudSpec{},
+					},
+					Version: semver.Semver{
+						Version: version,
+					},
+				},
+			},
+			ExpectedStatus: apiv1.ExternalCCMStatus{
+				ExternalCCM:          false,
+				ExternalCCMMigration: apiv1.ExternalCCMMigrationSupported,
+			},
+		},
+		{
+			Name: "scenario 4: CCM migration unsupported",
+			Datacenter: &kubermaticv1.Datacenter{
+				Spec: kubermaticv1.DatacenterSpec{Openstack: &kubermaticv1.DatacenterSpecOpenstack{}},
+			},
+			Cluster: &kubermaticv1.Cluster{
+				Spec: kubermaticv1.ClusterSpec{
+					Cloud: kubermaticv1.CloudSpec{
+						AWS: &kubermaticv1.AWSCloudSpec{},
```

Comment:
Tiny nit: I would replace this with a Fake or Kubeadm provider since AWS will support the CCM/CSI migration in the future.